lib/action_subscriber/publisher/async/in_memory_adapter.rb

```diff
@@ -0,0 +1,153 @@
+          else
+            NETWORK_ERRORS = [::Bunny::Exception, ::Timeout::Error, ::IOError].freeze
+          end
+
+          def initialize
+            @queue = ::Queue.new
+            create_and_supervise_consumer!
+          end
+
+          def push(message)
+            # Default of 1_000_000 messages.
+            if queue.size > ::ActionSubscriber.configuration.async_publisher_max_queue_size
+              # Drop Messages if the queue is full and we were configured to do so.
+              return if ::ActionSubscriber.configuration.async_publisher_drop_messages_when_queue_full
+
+              # By default we will raise an error to push the responsibility onto the caller.
+              fail UnableToPersistMessageError, "Queue is full, messages will be dropped."
+            end
+
+            queue.push(message)
+          end
+
+          def size
+            queue.size
+          end
+
+        private
+
+          def await_network_reconnect
+            sleep ::ActionSubscriber::RabbitConnection::NETWORK_RECOVERY_INTERVAL
+          end
+
+          def create_and_supervise_consumer!
+            @consumer = create_consumer
+            @supervisor = ::Thread.new do
+              loop do
+                unless consumer.alive?
+                  # Why might need to requeue the last message.
+                  queue.push(@current_message) if @current_message.present?
+                  consumer.kill
+                  @consumer = create_consumer
+                end
+
+                # Pause before checking the consumer again.
+                sleep supervisor_interval
+              end
+            end
+          end
+
+          def create_consumer
+            ::Thread.new do
+              loop do
+                # Write "current_message" so we can requeue should something happen to the consumer. I don't love this, but it's
+                # better than writing my own `#peek' method.
+                @current_message = message = queue.pop
+
+                begin
+                  ::ActionSubscriber::Publisher.publish(message.route, message.payload, message.exchange_name, message.options)
+
+                  # Reset
+                  @current_message = nil
+                rescue *NETWORK_ERRORS
+                  # Sleep because the connection is down.
+                  await_network_reconnect
+
+                  # Requeue and try again.
+                  queue.push(message)
+                rescue => unknown_error
+                  # Do not requeue the message because something else horrible happened.
+                  @current_message = nil
+
+                  # Log the error.
+                  logger.info unknown_error.class
+                  logger.info unknown_error.message
+                  logger.info unknown_error.backtrace.join("\n")
+
+                  # TODO: Find a way to bubble this out of the thread for logging purposes.
```

Comment:
One approach could be to use something like [Celluloid Actor Linking](https://github.com/celluloid/celluloid/wiki/Linking) and `trap_exit` to catch and notify the supervisor actor of the error.